"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[2613],{3905:(e,t,r)=>{r.r(t),r.d(t,{MDXContext:()=>s,MDXProvider:()=>c,mdx:()=>h,useMDXComponents:()=>p,withMDXComponents:()=>m});var n=r(67294);function a(e,t,r){return t in e?Object.defineProperty(e,t,{value:r,enumerable:!0,configurable:!0,writable:!0}):e[t]=r,e}function i(){return i=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e},i.apply(this,arguments)}function o(e,t){var r=Object.keys(e);if(Object.getOwnPropertySymbols){var n=Object.getOwnPropertySymbols(e);t&&(n=n.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),r.push.apply(r,n)}return r}function l(e){for(var t=1;t<arguments.length;t++){var r=null!=arguments[t]?arguments[t]:{};t%2?o(Object(r),!0).forEach((function(t){a(e,t,r[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(r)):o(Object(r)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(r,t))}))}return e}function d(e,t){if(null==e)return{};var r,n,a=function(e,t){if(null==e)return{};var r,n,a={},i=Object.keys(e);for(n=0;n<i.length;n++)r=i[n],t.indexOf(r)>=0||(a[r]=e[r]);return a}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(n=0;n<i.length;n++)r=i[n],t.indexOf(r)>=0||Object.prototype.propertyIsEnumerable.call(e,r)&&(a[r]=e[r])}return a}var s=n.createContext({}),m=function(e){return function(t){var r=p(t.components);return n.createElement(e,i({},t,{components:r}))}},p=function(e){var t=n.useContext(s),r=t;return e&&(r="function"==typeof e?e(t):l(l({},t),e)),r},c=function(e){var t=p(e.components);return n.createElement(s.Provider,{value:t},e.children)},f={inlineCode:"code",wrapper:function(e){var t=e.children;return n.createElement(n.Fragment,{},t)}},u=n.forwardRef((function(e,t){var r=e.components,a=e.mdxType,i=e.originalType,o=e.parentName,s=d(e,["components","mdxType","originalType","parentName"]),m=p(r),c=a,u=m["".concat(o,".").concat(c)]||m[c]||f[c]||i;return r?n.createElement(u,l(l({ref:t},s),{},{components:r})):n.createElement(u,l({ref:t},s))}));function h(e,t){var r=arguments,a=t&&t.mdxType;if("string"==typeof e||a){var i=r.length,o=new Array(i);o[0]=u;var l={};for(var d in t)hasOwnProperty.call(t,d)&&(l[d]=t[d]);l.originalType=e,l.mdxType="string"==typeof e?e:a,o[1]=l;for(var s=2;s<i;s++)o[s]=r[s];return n.createElement.apply(null,o)}return n.createElement.apply(null,r)}u.displayName="MDXCreateElement"},33490:(e,t,r)=>{r.r(t),r.d(t,{assets:()=>d,contentTitle:()=>o,default:()=>p,frontMatter:()=>i,metadata:()=>l,toc:()=>s});var n=r(83117),a=(r(67294),r(3905));const i={state:"draft"},o="Errors and Exceptions",l={unversionedId:"spec/definition/exception",id:"spec/definition/exception",title:"Errors and Exceptions",description:"How error classifications, functions qualifiers, and exceptions work together.",source:"@site/../doc/spec/definition/exception.md",sourceDirName:"spec/definition",slug:"/spec/definition/exception",permalink:"/fbthrift/docs/spec/definition/exception",draft:!1,editUrl:"https://github.com/facebook/fbthrift/blob/main/thrift/website/../doc/spec/definition/exception.md",tags:[],version:"current",frontMatter:{state:"draft"},sidebar:"sidebar",previous:{title:"Data Types",permalink:"/fbthrift/docs/spec/definition/data"},next:{title:"Interface Types",permalink:"/fbthrift/docs/spec/definition/interface"}},d={},s=[{value:"Exceptions",id:"exceptions",level:2},{value:"Error Classification",id:"error-classification",level:2},{value:"Kind",id:"kind",level:3},{value:"Fault Attribution",id:"fault-attribution",level:3},{value:"Error Safety",id:"error-safety",level:3},{value:"Example",id:"example",level:3},{value:"RPC Idempotency",id:"rpc-idempotency",level:2},{value:"Example",id:"example-1",level:3},{value:"Retryability",id:"retryability",level:2}],m={toc:s};function p(e){let{components:t,...r}=e;return(0,a.mdx)("wrapper",(0,n.Z)({},m,r,{components:t,mdxType:"MDXLayout"}),(0,a.mdx)("h1",{id:"errors-and-exceptions"},"Errors and Exceptions"),(0,a.mdx)("p",null,"How error classifications, functions qualifiers, and exceptions work together."),(0,a.mdx)("h2",{id:"exceptions"},"Exceptions"),(0,a.mdx)("p",null,"A ",(0,a.mdx)("strong",{parentName:"p"},"declared")," exception is an exception struct defined in IDL. Interfaces can declare that they might throw a declared exception using the ",(0,a.mdx)("inlineCode",{parentName:"p"},"throws")," keyword. Each method can specify a set of declared exceptions. Please refer to the ",(0,a.mdx)("a",{parentName:"p",href:"https://github.com/facebook/fbthrift/blob/main/thrift/lib/thrift/schema.thrift"},"schema")," for more detail and syntax."),(0,a.mdx)("p",null,"An ",(0,a.mdx)("strong",{parentName:"p"},"undeclared")," exception is thrown by Thrift methods that are wrapped into a generic ",(0,a.mdx)("a",{parentName:"p",href:"https://github.com/facebook/fbthrift/blob/main/thrift/lib/cpp/TApplicationException.h"},"TApplicationException")," or ",(0,a.mdx)("a",{parentName:"p",href:"https://github.com/facebook/fbthrift/blob/main/thrift/lib/cpp/transport/TTransportException.h"},"TTransportException")," and passed to the client. It is similar to throwing generic exceptions (e.g. ",(0,a.mdx)("inlineCode",{parentName:"p"},"Exception")," or ",(0,a.mdx)("inlineCode",{parentName:"p"},"std::exception"),") and is generally discouraged since it is too general to carry sufficient information. It is highly encouraged to avoid as much as possible throwing undeclared exceptions."),(0,a.mdx)("h2",{id:"error-classification"},"Error Classification"),(0,a.mdx)("h3",{id:"kind"},"Kind"),(0,a.mdx)("p",null,"Thrift Error can be classified into three categories: ",(0,a.mdx)("strong",{parentName:"p"},"Transient"),", ",(0,a.mdx)("strong",{parentName:"p"},"Stateful"),", and ",(0,a.mdx)("strong",{parentName:"p"},"Permanent"),"."),(0,a.mdx)("table",null,(0,a.mdx)("thead",{parentName:"table"},(0,a.mdx)("tr",{parentName:"thead"},(0,a.mdx)("th",{parentName:"tr",align:null},"Choices"),(0,a.mdx)("th",{parentName:"tr",align:null},"Source of Truth"),(0,a.mdx)("th",{parentName:"tr",align:null},"Meaning"),(0,a.mdx)("th",{parentName:"tr",align:null},"Retriable?"))),(0,a.mdx)("tbody",{parentName:"table"},(0,a.mdx)("tr",{parentName:"tbody"},(0,a.mdx)("td",{parentName:"tr",align:null},"Transient"),(0,a.mdx)("td",{parentName:"tr",align:null},"Exception definition in the server IDL"),(0,a.mdx)("td",{parentName:"tr",align:null},"Might succeed if retried immediately."),(0,a.mdx)("td",{parentName:"tr",align:null},"Maybe")),(0,a.mdx)("tr",{parentName:"tbody"},(0,a.mdx)("td",{parentName:"tr",align:null},"Stateful"),(0,a.mdx)("td",{parentName:"tr",align:null},"Exception definition in the server IDL"),(0,a.mdx)("td",{parentName:"tr",align:null},"State must be changed before any chance of success."),(0,a.mdx)("td",{parentName:"tr",align:null},"No")),(0,a.mdx)("tr",{parentName:"tbody"},(0,a.mdx)("td",{parentName:"tr",align:null},"Permanent"),(0,a.mdx)("td",{parentName:"tr",align:null},"Exception definition in the server IDL"),(0,a.mdx)("td",{parentName:"tr",align:null},"Can never succeed."),(0,a.mdx)("td",{parentName:"tr",align:null},"No")))),(0,a.mdx)("p",null,"Only a ",(0,a.mdx)("strong",{parentName:"p"},"transient")," error is a candidate for internal retries. A ",(0,a.mdx)("strong",{parentName:"p"},"stateful")," error requires application specific actions to be taken before a retry can succeed, for example, granting access to a server resource. A ",(0,a.mdx)("strong",{parentName:"p"},"permanent")," error fundamentally can never succeed. An error kind can be set using the keyword ",(0,a.mdx)("inlineCode",{parentName:"p"},"transient"),", ",(0,a.mdx)("inlineCode",{parentName:"p"},"stateful"),", and ",(0,a.mdx)("inlineCode",{parentName:"p"},"permanent"),". Please refer to the ",(0,a.mdx)("a",{parentName:"p",href:"https://github.com/facebook/fbthrift/blob/main/thrift/lib/thrift/schema.thrift"},"schema")," for more detail and syntax."),(0,a.mdx)("h3",{id:"fault-attribution"},"Fault Attribution"),(0,a.mdx)("p",null,"Attribution of fault for errors is classified by which party is at fault:"),(0,a.mdx)("table",null,(0,a.mdx)("thead",{parentName:"table"},(0,a.mdx)("tr",{parentName:"thead"},(0,a.mdx)("th",{parentName:"tr",align:null},"Choices"),(0,a.mdx)("th",{parentName:"tr",align:null},"Source of Truth"),(0,a.mdx)("th",{parentName:"tr",align:null},"Meaning"))),(0,a.mdx)("tbody",{parentName:"table"},(0,a.mdx)("tr",{parentName:"tbody"},(0,a.mdx)("td",{parentName:"tr",align:null},"Unspecified"),(0,a.mdx)("td",{parentName:"tr",align:null},"Exception definition in the server IDL"),(0,a.mdx)("td",{parentName:"tr",align:null},"The default. The server is assumed to be at fault for retry purposes.")),(0,a.mdx)("tr",{parentName:"tbody"},(0,a.mdx)("td",{parentName:"tr",align:null},"Server"),(0,a.mdx)("td",{parentName:"tr",align:null},"Exception definition in the server IDL"),(0,a.mdx)("td",{parentName:"tr",align:null},"The server is at fault.")),(0,a.mdx)("tr",{parentName:"tbody"},(0,a.mdx)("td",{parentName:"tr",align:null},"Client"),(0,a.mdx)("td",{parentName:"tr",align:null},"Exception definition in the server IDL"),(0,a.mdx)("td",{parentName:"tr",align:null},"The client is at fault.")))),(0,a.mdx)("p",null,"This is also typically statically associated with an Exception/ErrorCode, For example, it is built in to the HTTP error code space, where 4xx is a client error and 5xx is a server error. So, it will only be configurable in the IDL, and the server is used as the source of truth. An error fault attribution can be set using the keyword ",(0,a.mdx)("inlineCode",{parentName:"p"},"server")," and ",(0,a.mdx)("inlineCode",{parentName:"p"},"client"),". Please refer to the ",(0,a.mdx)("a",{parentName:"p",href:"https://github.com/facebook/fbthrift/blob/main/thrift/lib/thrift/schema.thrift"},"schema")," for more detail and syntax."),(0,a.mdx)("h3",{id:"error-safety"},"Error Safety"),(0,a.mdx)("table",null,(0,a.mdx)("thead",{parentName:"table"},(0,a.mdx)("tr",{parentName:"thead"},(0,a.mdx)("th",{parentName:"tr",align:null},"Choices"),(0,a.mdx)("th",{parentName:"tr",align:null},"Source of Truth"),(0,a.mdx)("th",{parentName:"tr",align:null},"Meaning"))),(0,a.mdx)("tbody",{parentName:"table"},(0,a.mdx)("tr",{parentName:"tbody"},(0,a.mdx)("td",{parentName:"tr",align:null},"Unspecified"),(0,a.mdx)("td",{parentName:"tr",align:null},"Exception definition in the server IDL"),(0,a.mdx)("td",{parentName:"tr",align:null},"The default, error is assumed to not be safe.")),(0,a.mdx)("tr",{parentName:"tbody"},(0,a.mdx)("td",{parentName:"tr",align:null},"Safe"),(0,a.mdx)("td",{parentName:"tr",align:null},"Exception definition in the server IDL"),(0,a.mdx)("td",{parentName:"tr",align:null},"Failed RPC is known to have no side-effects.")))),(0,a.mdx)("p",null,"A safe error guarantees that there are no significant side effects from having tried to process the request. For example, if a pre-condition check fails before the request was processed, that error is safe. It is always ok to retry a safe error. State and permanent errors are often statically associated safe. Transient errors are rarely safe. An error can be marked safe using the keyword ",(0,a.mdx)("inlineCode",{parentName:"p"},"safe"),". Please refer to the ",(0,a.mdx)("a",{parentName:"p",href:"https://github.com/facebook/fbthrift/blob/main/thrift/lib/thrift/schema.thrift"},"schema")," for more detail and syntax."),(0,a.mdx)("h3",{id:"example"},"Example"),(0,a.mdx)("pre",null,(0,a.mdx)("code",{parentName:"pre",className:"language-thrift"},"// Indicates the underlying machine holding the data could not be reached.\n//\n// For this error, the service is at fault, it is safe to retry, and a retry\n// might succeed if retried immediately.\nsafe transient server exception DataUnavailableError {...}\n\n// Indicates the client is sending data type that is not supported by the server.\n//\n// This error is safe to retry, but it will never succeed, and it originates from the\n// client.\nsafe permanent client exception UnSupportedType {...}\n\n// None of the qualifiers are required and may be ommitted, in which case the default for that category will be used.\n// Assumed to be not safe.\npermanent client exception UnSupportedType {...}\n")),(0,a.mdx)("h2",{id:"rpc-idempotency"},"RPC Idempotency"),(0,a.mdx)("p",null,(0,a.mdx)("strong",{parentName:"p"},"RPC idempotency")," determines the retryability of the RPC method using reserved function qualifier keywords ",(0,a.mdx)("inlineCode",{parentName:"p"},"readonly")," and ",(0,a.mdx)("inlineCode",{parentName:"p"},"idempotent"),". A ",(0,a.mdx)("em",{parentName:"p"},"readonly")," method has no meaningful side effects, so it can be called any time. A ",(0,a.mdx)("em",{parentName:"p"},"idempotent")," method can be immediately retried after a transient failure. Please refer to the ",(0,a.mdx)("a",{parentName:"p",href:"https://github.com/facebook/fbthrift/blob/main/thrift/lib/thrift/schema.thrift"},"schema")," and the ",(0,a.mdx)("a",{parentName:"p",href:"/fbthrift/docs/spec/definition/interface"},"interface definition")," for more detail and syntax."),(0,a.mdx)("h3",{id:"example-1"},"Example"),(0,a.mdx)("pre",null,(0,a.mdx)("code",{parentName:"pre",className:"language-thrift"},"service MyService {\n  // It does not change the service, so it is always safe to rety (even after a success).\n  readonly Data getData(...);\n  // Safe to rety immeidately after a transient failure.\n  idempotent void deleteData(...);\n  // Not safe to retry unless the error is 'safe'.\n  void updateData(...);\n}\n")),(0,a.mdx)("h2",{id:"retryability"},"Retryability"),(0,a.mdx)("p",null,"RPC Retry is enabled by default. Based on the error classification and RPC idempotency, the retry policy can be defined as the following:"),(0,a.mdx)("pre",null,(0,a.mdx)("code",{parentName:"pre"},"Retryable = Error.Transient &&\n    (RPC.Idempotent || RPC.Readonly || Error.Safe)\n")),(0,a.mdx)("p",null,"Note that, if the error kind is unspecified, it is assumed to be transient, and if the error is not known to be safe, it is assumed to not be safe. Thus the \u2018default\u2019 behavior is to only retry if the RPC is idempotent."),(0,a.mdx)("p",null,"Unfortunately, transient errors, which are the only ones worth retrying, are rarely known to always be safe. However, most transient errors are produced by Thrift itself and known internally to be safe."))}p.isMDXComponent=!0}}]);